File: src/sacramento.c
#include <R.h>
#include "sacramento.h"

/* Trivial wrapper function */
/* Based on code from University of Arizona MOSCEM project */
/* Simplified by Felix Andrews <felix@nfrac.org> 2010-02-01 */

void sma_sac(double *P, double *E, int *n,
	     double *xpar, double *etmult, 
	     double *dt, double *U)
{
    struct SMA sma;
    struct FSUM1 fsum1;

    /* ASSIGN MODEL SPECIFIC PARAMETERS TO PARAMETERS STRUCT */
    sma.uztwm = xpar[0];
    sma.uzfwm = xpar[1];
    sma.uzk   = xpar[2];
    sma.pctim = xpar[3];
    sma.adimp = xpar[4];
    sma.riva  = 0.0;
    sma.zperc = xpar[5];
    sma.rexp  = xpar[6];
    sma.lztwm = xpar[7];
    sma.lzfsm = xpar[8];
    sma.lzfpm = xpar[9];
    sma.lzsk  = xpar[10];
    sma.lzpk  = xpar[11];
    sma.pfree = xpar[12];
    sma.rserv = 0.3;
    sma.side  = 0.0;
    sma.pxmlt = 1.0;

    sma.uztwc = 0.5 * sma.uztwm;
    sma.uzfwc = 0.5 * sma.uzfwm;
    sma.lztwc = 0.5 * sma.lztwm;
    sma.lzfsc = 0.5 * sma.lzfsm;
    sma.lzfpc = 0.5 * sma.lzfpm;
    sma.adimc = 0.5 * (sma.uztwm + sma.lztwm);

    /* SET SOME INITIAL VALUES TO ZERO */
    fsum1.srot=fsum1.simpvt=fsum1.srodt=fsum1.srost=0.;
    fsum1.sintft=fsum1.sgwfp=fsum1.sgwfs=fsum1.srecht=fsum1.sett=0.;
    fsum1.se1=fsum1.se3=fsum1.se4=fsum1.se5=0.;


    /* SET EVAPOTRANSPIRATION DISTRIBUTION */
    sma.epdist = *etmult;

    /* DT IS THE LENGTH OF EACH TIME INTERVAL IN DAYS */
    /* DT IS USED TO CALCULATE dinc IN fland1 */
    sma.dt = *dt;  /*  */

    for (int t = 0; t < *n; t++) {

	/* ASSIGN AND ADJUST PET VALUE */
	sma.ep =  E[t] * sma.epdist;

	/* ASSIGN AND ADJUST PRECIPITATION VALUE */
	sma.pxv = P[t] * sma.pxmlt;

	/* PERFORM SOIL MOISTURE ACCOUNTING OPERATIONS  */
	fland1(&sma,&fsum1);

	/* SET TOTAL CHANNEL INFLOW EQUAL TO THE EXCESS AT THE END  */
	/* OF EACH TIME PERIOD */
	U[t] = sma.tlci;
    }
}

File: R/sacramento.R
## hydromad: Hydrological Modelling and Analysis of Data
##
## Copyright (c) Felix Andrews <felix@nfrac.org>
##

## Sacramento Soil Moisture Accounting model.
## Developed by the US National Weather Service.
sacramento.sim <-
    function(DATA,
             uztwm, uzfwm, uzk, pctim, adimp, zperc, rexp,
             lztwm, lzfsm, lzfpm, lzsk, lzpk, pfree,
             etmult = 1, dt = 1,
             uztwc_0=0.5,uzfwc_0=0.5,
             lztwc_0=0.5,lzfsc_0=0.5,lzfpc_0=0.5,
             adimc_0=0.5,min_ninc=20,
             return_state = FALSE)
{
    stopifnot(c("P","E") %in% colnames(DATA))
    ## check values
    stopifnot(uztwm >= 0)
    stopifnot(uzfwm >= 0)
    stopifnot(uzk >= 0)
    stopifnot(0 <= pctim && pctim <= 1)
    stopifnot(adimp >= 0)
    stopifnot(zperc >= 0)
    stopifnot(lztwm >= 0)
    stopifnot(lzfsm >= 0)
    stopifnot(lzfpm >= 0)
    stopifnot(lzsk >= 0)
    stopifnot(lzpk >= 0)
    stopifnot(pfree >= 0)
    stopifnot(etmult >= 0)
    stopifnot(dt >= 0)

    xpar <-
        c(uztwm, uzfwm, uzk, pctim, adimp, zperc, rexp,
          lztwm, lzfsm, lzfpm, lzsk, lzpk, pfree)

    P <- DATA[,"P"]
    E <- DATA[,"E"]
    ## skip over missing values
    bad <- is.na(P) | is.na(E)
    P[bad] <- 0
    E[bad] <- 0
    ## NOTE: there is no pure.R.code implementation
    if (return_state) {
        states <- .C(sma_sac_state,
                     as.double(P),
                     as.double(E),
                     as.integer(NROW(DATA)),
                     as.double(xpar),
                     as.double(etmult),
                     as.double(dt),
                     U = double(NROW(DATA)),
                     uztwc = double(NROW(DATA)),
                     uzfwc = double(NROW(DATA)),
                     lztwc = double(NROW(DATA)),
                     lzfsc = double(NROW(DATA)),
                     lzfpc = double(NROW(DATA)),
                     adimc = double(NROW(DATA)),
                     sett = double(NROW(DATA)),
                     se1 = double(NROW(DATA)),
                     se3 = double(NROW(DATA)),
                     se4 = double(NROW(DATA)),
                     se5 = double(NROW(DATA)),
                     roimp = double(NROW(DATA)),
                     sdro = double(NROW(DATA)),
                     ssur = double(NROW(DATA)),
                     sif = double(NROW(DATA)),
                     bfp = double(NROW(DATA)),
                     bfs = double(NROW(DATA)),
                     bfcc = double(NROW(DATA)),
                     as.double(uztwc_0),as.double(uzfwc_0),
                     as.double(lztwc_0),as.double(lzfsc_0),as.double(lzfpc_0),
                     as.double(adimc_0),
                     as.integer(min_ninc),
                     NAOK = FALSE, DUP = FALSE, PACKAGE="hydromad")
        for(i in 7:25) attributes(states[[i]]) <- attributes(P)
        ans <- do.call(cbind,states[7:25])
        return(ans)
    } else {
        U <- .C(sma_sac,
                as.double(P),
                as.double(E),
                as.integer(NROW(DATA)),
                as.double(xpar),
                as.double(etmult),
                as.double(dt),
                U = double(NROW(DATA)),
                as.double(uztwc_0),as.double(uzfwc_0),
                as.double(lztwc_0),as.double(lzfsc_0),as.double(lzfpc_0),
                as.double(adimc_0),
                as.integer(min_ninc),
                NAOK = FALSE, DUP = FALSE, PACKAGE="hydromad")$U
        ## make it a time series object again
        attributes(U) <- attributes(P)
        ## re-insert missing values
        U[bad] <- NA
        return(U)
    }
}

sacramento.ranges <- function()
    list(uztwm = c(1, 150),
         uzfwm = c(1, 150),
         uzk = c(0.1, 0.5),
         pctim = c(0.000001, 0.1),
         adimp = c(0, 0.4),
         zperc = c(1, 250),
         rexp = c(0, 5),
         lztwm = c(1, 500),
         lzfsm = c(1, 1000),
         lzfpm = c(1, 1000),
         lzsk = c(0.01, 0.25),
         lzpk = c(0.0001, 0.25),
         pfree = c(0, 0.6))
